src/Compiler/MacroProcessor.cpp

```diff
@@ -9,15 +13,82 @@ namespace Ark::internal
     MacroProcessor::MacroProcessor(unsigned debug, uint16_t options) noexcept :
         m_debug(debug), m_options(options)
     {
-        m_trueNode = Node("true");
-        m_falseNode = Node("false");
-        m_nilNode = Node("nil");
-        m_listNode = Node("list");
-
-        m_trueNode.setNodeType(NodeType::Symbol);
-        m_falseNode.setNodeType(NodeType::Symbol);
-        m_nilNode.setNodeType(NodeType::Symbol);
-        m_listNode.setNodeType(NodeType::Symbol);
+        // initialize default Nodes
+        Node::init();
+
+        const auto& apply_to_func 
+        { 
+            [this](const std::unordered_map<std::string, Node>& map, Node& target, Node* parent)
+            {
+                return this->apply_to(map, target, parent);
+            }
+        };
+ 
+        const auto& func_isTruthy 
+        { 
+            [this](const Node& node)
+            {
+                return this->isTruthy(node);
+            }
+        };
+        const auto& func_evaluate 
+        {
+            [this](Node& node, bool is_not_body = false)
+            {
+                return this->evaluate(node, is_not_body);
+            }
+        };
+        const auto& func_find_nearest_macro {
+            [this](const std::string& name)
+            {
+                return this->find_nearest_macro(name);
+            }
+        };
+
+        const auto& func_registerMacro
+        {
+            [this](Node &node)
+            {
+            this->registerMacro(node);
+            }
+        };
+
+        const auto& func_processingError 
+        {
+            [this](const std::string& message, const Node& node)
+            {
+            this->throwMacroProcessingError(message, node);
+            }
+        };
+
+        const auto& func_execute 
+        {
+            [this](Node& node)
+            {
+                return this->execute(node);
+            }
+        };
+
+      
+        std::vector<std::shared_ptr<MacroExecutor>> executors = 
+        {
+            std::make_shared<SymbolExecutor>(),
+            std::make_shared<ConditionalExecutor>(),
+            std::make_shared<ListExecutor>(),
+        };
+
+        executor_pipeline = std::make_unique<MacroExecutorPipeline>(
+            func_find_nearest_macro,
+            func_registerMacro,
+            func_isTruthy,
+            func_evaluate,
+            apply_to_func,
+            func_processingError,
+            func_execute,
+            executors
+        );
```

Comment:
Maybe a bit late but I had an idea: instead of passing std::functions capturing `this` to the macro pipeline, how about giving it `this` (which is of type `MacroProcessor*`) so that the pipeline members will directly use the pointer and call the methods?

It will need to make them friend with the MacroProcessor however, but a single `friend class ABC;` in the header will do the work. As you can see in this example: http://coliru.stacked-crooked.com/a/2d7fce5b94bfdd51, you will only need to declare the base class as a friend of the MacroProcessor, and all the subclasses will automatically be friends with the MacroProcessor

Reply:
This was my first idea but I had trouble dealing with the circular dependency between MacroProcessor and MacroExecutor. Do I forward declare them both in each other's header files?
I like this way better. I'm open for suggestions on how to handle this

Reply:
Since you only need a pointer, a forward declaration in the headers is fine

To avoid circular deps and incomplete types, you will need to #include <> the macro processor in the .cpp only

Reply:
This will reduce maintenance cost. I'll do this first because it might cancel out the other code changes. Thanks :D